server/src/main/java/io/crate/execution/engine/export/FileWriterCountCollector.java

```diff
@@ -57,49 +57,53 @@
  * and returns a count representing the number of written rows
  */
 public class FileWriterCountCollector implements Collector<Row, long[], Iterable<Row>> {
 
     private static final byte NEW_LINE = (byte) '\n';
 
+    private final Executor executor;
     private final Iterable<CollectExpression<Row, ?>> collectExpressions;
     private final List<Input<?>> inputs;
+    private final URI uri;
+    private final FileOutput fileOutput;
     private final Map<String, Object> overwrites;
+    private final WriterProjection.CompressionType compressionType;
     @Nullable
     private final List<String> outputNames;
     private final WriterProjection.OutputFormat outputFormat;
-    private Output output;
 
     private final RowWriter rowWriter;
 
     FileWriterCountCollector(Executor executor,
-                             String uri,
+                             String uriStr,
                              @Nullable WriterProjection.CompressionType compressionType,
                              @Nullable List<Input<?>> inputs,
                              Iterable<CollectExpression<Row, ?>> collectExpressions,
                              Map<ColumnIdent, Object> overwrites,
                              @Nullable List<String> outputNames,
-                             WriterProjection.OutputFormat outputFormat) {
+                             WriterProjection.OutputFormat outputFormat,
+                             Map<String, FileOutputFactory> fileOutputFactories) {
+        this.executor = executor;
         this.collectExpressions = collectExpressions;
         this.inputs = inputs;
         this.overwrites = toNestedStringObjectMap(overwrites);
+        this.compressionType = compressionType;
         this.outputNames = outputNames;
         this.outputFormat = outputFormat;
-        URI uri1;
         try {
-            uri1 = new URI(uri);
+            uri = new URI(uriStr);
         } catch (URISyntaxException e) {
-            throw new SQLParseException(String.format(Locale.ENGLISH, "Invalid uri '%s'", uri), e);
+            throw new SQLParseException(String.format(Locale.ENGLISH, "Invalid uri '%s'", uriStr), e);
         }
-        if (uri1.getScheme() == null || uri1.getScheme().equals("file")) {
-            this.output = new OutputFile(uri1, compressionType);
-        } else if (uri1.getScheme().equalsIgnoreCase("s3")) {
-            this.output = new OutputS3(executor, uri1, compressionType);
-        } else {
-            throw new UnsupportedFeatureException(String.format(Locale.ENGLISH, "Unknown scheme '%s'", uri1.getScheme()));
+        String scheme = uri.getScheme();
+        scheme = (scheme == null) ? LocalFsFileOutputFactory.NAME : scheme;
+        if (!fileOutputFactories.containsKey(scheme)) {
+            throw new UnsupportedFeatureException(String.format(Locale.ENGLISH, "Unknown scheme '%s'", scheme));
         }
+        fileOutput = fileOutputFactories.get(scheme).create();
```

Comment:
Could turn this into a `.get()` plus `!= null check. Otherwise both `containsKey` and `get` have to do the work of looking up the item.